File: app/src/main/java/com/kira/kdownloader/ui/HomeViewModel.java
package com.kira.kdownloader.ui;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.kira.kdownloader.engine.DownloaderRepository;
import com.kira.kdownloader.engine.MediaInfo;
import com.kira.kdownloader.util.AppExecutors;
import com.kira.kdownloader.util.UrlExtractor;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.Future;
import java.util.concurrent.atomic.AtomicInteger;

public final class HomeViewModel extends AndroidViewModel {
    private static final long CACHE_TTL_MS = 5L * 60L * 1000L;
    private static final int MAX_CACHE_ENTRIES = 20;

    private final DownloaderRepository repository;
    private final MutableLiveData<HomeUiState> state =
            new MutableLiveData<>(HomeUiState.Idle.INSTANCE);
    private final LinkedHashMap<String, CachedInfo> infoCache = new LinkedHashMap<>();
    private final AtomicInteger fetchGeneration = new AtomicInteger();
    private Future<?> fetchFuture;
    private boolean warmed;

    public HomeViewModel(@NonNull Application application) {
        super(application);
        repository = new DownloaderRepository(application);
    }

    public LiveData<HomeUiState> getState() {
        return state;
    }

    public synchronized void warmUp() {
        if (warmed) return;
        warmed = true;
        AppExecutors.io().execute(() -> {
            try {
                repository.init();
            } catch (Throwable ignored) {
                // Fetch reports initialization errors to the UI when the user actually requests it.
            }
        });
    }

    public synchronized void fetch(String text) {
        String url = UrlExtractor.fromText(text);
        if (url.isEmpty()) {
            state.setValue(new HomeUiState.Error("Enter a media URL first"));
            return;
        }

        MediaInfo cached = cachedInfo(url);
        if (cached != null) {
            cancelFetch();
            state.setValue(new HomeUiState.Loaded(url, cached));
            return;
        }

        cancelFetch();
        int generation = fetchGeneration.incrementAndGet();
        state.setValue(HomeUiState.Loading.INSTANCE);
        fetchFuture = AppExecutors.io().submit(() -> {
            try {
                repository.init();
                MediaInfo info = repository.fetchInfo(url);
                if (generation != fetchGeneration.get() || Thread.currentThread().isInterrupted()) return;
                putCache(url, info);
                state.postValue(new HomeUiState.Loaded(url, info));
            } catch (Throwable error) {
                if (generation != fetchGeneration.get() || Thread.currentThread().isInterrupted()) return;
                String message = error.getMessage();
                state.postValue(new HomeUiState.Error(
                        message == null || message.trim().isEmpty()
                                ? "Failed to fetch media info" : message));
            }
        });
    }

    public synchronized void reset() {
        cancelFetch();
        state.setValue(HomeUiState.Idle.INSTANCE);
    }

    private synchronized void cancelFetch() {
        fetchGeneration.incrementAndGet();
        if (fetchFuture != null) fetchFuture.cancel(true);
        fetchFuture = null;
    }

    private synchronized MediaInfo cachedInfo(String url) {
        CachedInfo cached = infoCache.get(url);
        if (cached == null || System.currentTimeMillis() - cached.atMs >= CACHE_TTL_MS) return null;
        return cached.info;
    }

    private synchronized void putCache(String url, MediaInfo info) {
        infoCache.put(url, new CachedInfo(info, System.currentTimeMillis()));
        while (infoCache.size() > MAX_CACHE_ENTRIES) {
            String first = infoCache.keySet().iterator().next();
            infoCache.remove(first);
        }
    }

    @Override protected void onCleared() {
        cancelFetch();
        repository.close();
        super.onCleared();
    }

    private static final class CachedInfo {
        private final MediaInfo info;
        private final long atMs;

        private CachedInfo(MediaInfo info, long atMs) {
            this.info = info;
            this.atMs = atMs;
        }
    }
}

File: app/src/main/java/com/kira/kdownloader/ui/HomeViewModel.kt
package com.kira.kdownloader.ui

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.kira.kdownloader.engine.DownloaderRepository
import com.kira.kdownloader.engine.MediaInfo
import com.kira.kdownloader.util.RecentUrls
import com.kira.kdownloader.util.UrlExtractor
import kotlinx.coroutines.Job
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

sealed interface HomeUiState {
    data object Idle : HomeUiState
    data object Loading : HomeUiState
    data class Loaded(val sourceUrl: String, val info: MediaInfo) : HomeUiState
    data class Error(val message: String) : HomeUiState
}

class HomeViewModel(application: Application) : AndroidViewModel(application) {
    private val repository = DownloaderRepository(application)
    private val mutableState = MutableStateFlow<HomeUiState>(HomeUiState.Idle)
    private var fetchJob: Job? = null
    private var warmed = false

    // Short-lived cache of fetched formats, keyed by normalized URL. The TTL keeps it useful for
    // immediate retries / re-picks while avoiding serving direct media URLs after they've expired.
    private data class CachedInfo(val info: MediaInfo, val atMs: Long)
    private val infoCache = LinkedHashMap<String, CachedInfo>()

    val state: StateFlow<HomeUiState> = mutableState.asStateFlow()

    /** Initialize the engine ahead of the first fetch so its one-time cost is already paid. */
    fun warmUp() {
        if (warmed) return
        warmed = true
        viewModelScope.launch(Dispatchers.IO) { runCatching { repository.init() } }
    }

    fun fetch(url: String) {
        val normalizedUrl = UrlExtractor.fromText(url)
        if (normalizedUrl.isEmpty()) {
            mutableState.value = HomeUiState.Error("Enter a media URL first")
            return
        }

        cachedInfo(normalizedUrl)?.let { info ->
            fetchJob?.cancel()
            mutableState.value = HomeUiState.Loaded(sourceUrl = normalizedUrl, info = info)
            return
        }

        fetchJob?.cancel()
        fetchJob = viewModelScope.launch {
            mutableState.value = HomeUiState.Loading
            try {
                val info = withContext(Dispatchers.IO) {
                    repository.init()
                    repository.fetchInfo(normalizedUrl)
                }
                RecentUrls.add(getApplication<Application>(), normalizedUrl)
                putCache(normalizedUrl, info)
                mutableState.value = HomeUiState.Loaded(
                    sourceUrl = normalizedUrl,
                    info = info,
                )
            } catch (error: Throwable) {
                mutableState.value = HomeUiState.Error(
                    error.message?.takeIf(String::isNotBlank) ?: "Failed to fetch media info",
                )
            }
        }
    }

    private fun cachedInfo(normalizedUrl: String): MediaInfo? =
        infoCache[normalizedUrl]
            ?.takeIf { System.currentTimeMillis() - it.atMs < CACHE_TTL_MS }
            ?.info

    private fun putCache(normalizedUrl: String, info: MediaInfo) {
        infoCache[normalizedUrl] = CachedInfo(info, System.currentTimeMillis())
        while (infoCache.size > MAX_CACHE_ENTRIES) {
            infoCache.remove(infoCache.keys.first())
        }
    }

    override fun onCleared() {
        repository.close()
        super.onCleared()
    }

    private companion object {
        const val CACHE_TTL_MS = 5 * 60_000L
        const val MAX_CACHE_ENTRIES = 20
    }
}
